# R methylation workflow
A pipeline to analyze methylation array data

  DNA methylation is an epigenetic mechanism essential for the regulation of gene expression and cell differentiation, but its abnormal dysregulation may have substantial consequences, such as the development of cancer. Lymphomas are a group of blood malignancies that develop from B, T or NK cells, featuring a wide range of heterogeneous subtypes, whose diagnosis and treatment is difficult, mainly because of the lack of knowledge about their molecular pathology. The goal of this workflow is the study of differential methylation between Diffused Large B-cell Lympomaâ€™s (DLBCL) subtypes and in comparison to healthy controls in order to identify genes that show significant differential methylation.
  In particular, a publicly available [dataset](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-2926) from the EBI (European Bioinformatics Institute) database was analysed. The analysis was performed in the R programming language using packages from the Bioconductor platform. 
An existing [workflow](https://www.bioconductor.org/packages/release/workflows/vignettes/methylationArrayAnalysis/inst/doc/methylationArrayAnalysis.html) was employed for the analysis of the methylation data, with certain modifications to fit the dataset. The steps of the analysis where a) preprocessing of the data including quality control, normalization and filtering, b) identification of statistically significant differentially methylated CpGs and genomic regions and c) gene set enrichment analysis using terms and pathways from the databases [Gene Ontology](http://geneontology.org/) (GO), [Kyoto Encyclopedia for Genes and Genomes]9https://www.genome.jp/kegg/) (KEGG) and [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/).
